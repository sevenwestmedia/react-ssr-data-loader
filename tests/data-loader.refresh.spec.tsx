import { ComponentFixture } from './helpers/component-fixture'
import { DifferentKeysDataComponentFixture } from './helpers/different-keys-data-component-fixture'
import Adapter from 'enzyme-adapter-react-16'
import { configure } from 'enzyme'
import { act } from 'react-dom/test-utils'

configure({ adapter: new Adapter() })

describe('data-loader', () => {
    it('can refresh data', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Test' })
        })

        sut.assertState()

        act(() => sut.refreshData())

        sut.assertState()

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Test2' })
        })

        sut.assertState()
    })

    it('calling refresh when already refreshing ignores the action', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Test' })
            sut.refreshData()
            sut.refreshData()
        })

        sut.assertState()
    })

    it('refresh action is called with the correct context', async () => {
        const sut = new DifferentKeysDataComponentFixture(
            undefined,
            'testKey',
            'testKey2',
            false,
            false,
        )

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Test1_1' })
            await sut.testDataPromise2.resolve({ result: 'Test2_1' })
            sut.refreshData2()
        })

        expect(sut.loadDataCount1).toBe(1)
        expect(sut.loadDataCount2).toBe(2)
    })
})
