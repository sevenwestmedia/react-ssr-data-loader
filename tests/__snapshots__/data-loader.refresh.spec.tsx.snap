// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-loader calling refresh when already refreshing ignores the action 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 3,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader can refresh data 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader can refresh data 2`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader can refresh data 3`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test2",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test2",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;
